### PR TITLE
Adding ability for deid to support deid-provided functions

### LINKED ISSUE #207: Add deid provided functions
Discussion here: https://github.com/pydicom/deid/pull/203#issuecomment-1071870150

@moloney did you say you wanted to tackle this, or would you like someone else to take a first shot?

### PR DESCRIPTION
**Pull request by vsoch**
As of this version, a user can specify a value as a deid_func:<name> meaning that we use a deid provided function. This is useful for providing some general uid and a customized jitter function, along with other functions that users might want to add. With this change I have provided docs and a contributing section for how to do this, along with running pyflakes on
the tests.

This will close #207 and will assist with #206 

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

### CODE REVIEW

**Comment by vsoch** [comment]
Ping @moloney and @wetzelj !

**Comment by moloney** [comment]
Sorry for dropping off on this, I am taking a look now.  One immediate comment:  why not use the pydicom "generate_uid" function as I had done previously?  It can replace all the code in "deid/dicom/actions/uids.py", and it can provide purely random UIDs as you are doing in your code or it can provide a consistent mapping (same input UID produces same anonymized UID) so that study / series UIDs remain consistent after anonymization without any coordination between the processing of the separate files. This latter functionality is critical for many use cases.

Thanks for the hard work here!

**Comment by vsoch** [comment]
@moloney that's a fantastic idea! I 100% agree, and I feel silly that I didn't know about it. Can you suggest a name / args etc of how it should look?

**Comment by moloney** [comment]
There are just two optional args, a prefix (defaults to the pydicom root prefix) and an optional list of things to use as entropy when generating the UID (if nothing is provided it just uses system randomness).  If you pass the original UID in as the only entropy source you will get a stable result across all files (same input results in same output).

Docs are here: https://pydicom.github.io/pydicom/dev/reference/generated/pydicom.uid.generate_uid.html

So I guess the deid func should just take two optional args, a prefix and a toggle to enable / disable stable remapping (i.e. passing in the original UID) which I think should default to being turned on.

**Comment by vsoch** [comment]
@moloney for the remapping bit - does that mean passing the original value as the prefix or in the list of entopy items (it sounds like the second for the stability?) I just want to double check before writing it!

**Comment by vsoch** [comment]
okay I think I have it figured out! I'll have an update to the PR for you after work today.

**Comment by vsoch** [comment]
@moloney all set! https://github.com/pydicom/deid/pull/208/commits/bca9596ea2766bac7cb9838a9b04f265156d3d5e

File: docs/_docs/contributing/code.md [hunk 1]
```diff
@@ -12,3 +12,55 @@ Generally, for code contribution you should:
 4. Open a pull request against the master branch.
 
 See the repository `CONTRIBUTING.md` for these same details.
+
+## Contributing a Custom Function
+
+Deid ships (as of version 0.2.3) with deid-provided functions that can be used in
+header parsing. To contribute a recipe you should do the following:
```
Comment (wetzelj) [inline docs/_docs/contributing/code.md:19]: Trivial comment - but this should probably read: "To contribute a **custom function**..."

@vsoch - Sorry for my delay looking at this and responding. I've continued to be pulled in other areas so haven't had much time to spend on deid...

Comment (vsoch) [reply docs/_docs/contributing/code.md:19]: Totally no worries! I'll be around when you are ready :)

File: docs/_docs/examples/func-replace.md [hunk 2]
```diff
@@ -10,6 +10,8 @@ change field with a value derived from a function. This example
 was derived based on a prompt in [this pull request](https://github.com/pydicom/contrib-pydicom/pull/14).
 If you are interested in the code for this example, it's available
 [here](https://github.com/pydicom/deid/tree/master/examples/dicom/header-manipulation).
+If you are interested in the functions provided by deid (and you don't want to write your
+own function) see [this documentation](https://pydicom.github.io/deid/user-docs/recipe-groups/).
```
Comment (wetzelj) [inline docs/_docs/examples/func-replace.md:14]: The link on "this documentation" is currently referencing recipe-groups, this should be a link to recipe-funcs.

**Review by wetzelj — COMMENTED** [review]
I've looked through everything in this PR and commented on a couple trivial things. I won't be able to pull this in do my app to do any testing for a bit, but honestly, I think these changes are great!  I like the idea of having DEID provided custom functions - and we'll have a few to contribute once we have an opportunity to clean them up for submission.

File: docs/_docs/examples/func-replace.md (hunk 2 above)
Comment (vsoch) [reply docs/_docs/examples/func-replace.md:14]: That's called me failing at copy paste :laughing:

Comment (vsoch) [reply docs/_docs/examples/func-replace.md]: Fixed.

**Comment by vsoch** [comment]
heads up y'all! I'm going to merge this end of day - I think it looks fairly good and is an improvement to deid and we've had overall positive review.

**Comment by wetzelj** [comment]
Sounds good to me!

**Pull request merged**